qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,146 @@
+        numa_cpu_cli = []
+        tmp = {}
+        for numa_cpu in numa_cpus:
+            numa_cpu_params = params.object_params(numa_cpu)
+            nodeid = numa_cpu_params["numa_cpu_nodeid"]
+            socket = numa_cpu_params.get("numa_cpu_socketid")
+            die = numa_cpu_params.get("numa_cpu_dieid")
+            core = numa_cpu_params.get("numa_cpu_coreid")
+            thread = numa_cpu_params.get("numa_cpu_threadid")
+            cpu_list = convert_cpu_topology_to_ids(socket, die, core, thread)
+            if nodeid in tmp.keys():
+                tmp[nodeid] += cpu_list
+            else:
+                tmp[nodeid] = cpu_list
+        for item in sorted(tmp.items(), key=lambda item: item[0]):
+            numa_cpu_cli.append(set(sorted(item[1])))
+        return numa_cpu_cli
+
+    def get_hotpluggable_cpus():
+        """
+        Get the cpu id list for each node with the output of "query-hotpluggable-cpus"
+        """
+        error_context.context("Get the hotpluggable cpus", logging.info)
+        specified_cpus = []
+        tmp = {}
+        out = vm.monitor.info("hotpluggable-cpus")
+        for vcpu_info in out:
+            vcpus_count = vcpu_info["vcpus-count"]
+            vcpu_info = vcpu_info["props"]
+            nodeid = vcpu_info.get("node-id")
+            socket = vcpu_info.get("socket-id")
+            die = vcpu_info.get("die-id")
+            core = vcpu_info.get("core-id")
+            thread = vcpu_info.get("thread-id")
+            if nodeid is not None:
+                cpu_list = convert_cpu_topology_to_ids(socket, die, core, thread)
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += cpu_list
+                else:
+                    tmp[nodeid] = cpu_list
+        for item in sorted(tmp.items(), key=lambda item: item[0]):
+            specified_cpus.append(set(sorted(item[1])))
+        return specified_cpus
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    os_type = params["os_type"]
+    vm_arch = params["vm_arch_name"]
+
+    vcpu_threads = params.get_numeric('vcpu_threads')
+    if vm_arch in ('x86_64', 'i386'):
+        vcpu_sockets = params.get_numeric('vcpu_sockets')
+        vcpu_dies = params.get_numeric('vcpu_dies')
+        vcpu_cores = params.get_numeric('vcpu_cores')
+
+        socket_weight = vcpu_dies * vcpu_cores * vcpu_threads
+        die_weight = vcpu_cores * vcpu_threads
+        core_weight = vcpu_threads
+        thread_weight = 1
+
+    numa_cpu_cli = numa_cpu_cli()
+    logging.info("Expected cpus from qemu cli: %s" % numa_cpu_cli)
+
+    specified_cpus = get_hotpluggable_cpus()
+    logging.info("[Hotpluggable]Specified cpus: %s" % specified_cpus)
+    numa_cpu_monitor = [item[1] for item in vm.monitor.info_numa()]
+    logging.info("[Monitor]numa cpus: %s" % numa_cpu_monitor)
+    if os_type == 'linux':
+        # Get numa cpus in guest os, only for Linux
+        numa_cpu_guest = numa_cpu_guest()
```

Comment:
Did you notice there are some duplicate lines in debug.log? Seems some cmds are sent more than once.

Reply:
> Did you notice there are some duplicate lines in debug.log? Seems some cmds are sent more than once.

Yes, check if file exist is one, and get the file info is another one. You can refer:
https://github.com/avocado-framework/avocado-vt/blob/a0f6c7eb548167a990ee30b1acdee97560c741d9/virttest/utils_misc.py#L1529

Reply:
I see. Except that, a few lines of 'DEBUG| Sending command: numactl --hardware'. But it has nothing to do with this patch but with class NumaInfo. We can put it aside for now.